ast/compile_test.go

```diff
@@ -3138,6 +3138,412 @@ func assertErrors(t *testing.T, actual Errors, expected Errors, assertLocation b
+				},
+			},
+		},
+		{
+			note: "var shadows data (multiple modules)",
+			modules: map[string]string{
+				"policy1.rego": `package test
+					import future.strict
+					p if {
+						data := 1
+						data == 1
+					}`,
+				"policy2.rego": `package test
+					import future.strict
+					q if {
+						data := 1
+						data == 1
+					}`,
+			},
+			expectedErrors: Errors{
+				&Error{
+					Message:  "variables must not shadow data (use a different variable name)",
+					Location: &Location{Text: []byte("data := 1"), File: "policy1.rego", Row: 4, Col: 7},
+				},
+				&Error{
+					Message:  "variables must not shadow data (use a different variable name)",
+					Location: &Location{Text: []byte("data := 1"), File: "policy2.rego", Row: 4, Col: 7},
+				},
+			},
+		},
+		{
+			note: "var shadows data (multiple modules, not all strict)",
+			modules: map[string]string{
+				"policy1.rego": `package test
+					import future.keywords.if
+					p if {
+						data := 1
+						data == 1
+					}`,
+				"policy2.rego": `package test
+					import future.strict
+					q if {
+						data := 1
+						data == 1
+					}`,
+			},
+			expectedErrors: Errors{
+				&Error{
+					Message:  "variables must not shadow data (use a different variable name)",
+					Location: &Location{Text: []byte("data := 1"), File: "policy2.rego", Row: 4, Col: 7},
+				},
+			},
+		},
+		// rule shadowing
+		{
+			note: "rule shadows input",
+			modules: map[string]string{
+				"policy.rego": `package test
+					import future.strict
+					input := 1`,
+			},
+			expectedErrors: Errors{
+				&Error{
+					Message:  "rules must not shadow input (use a different rule name)",
+					Location: &Location{Text: []byte("input := 1"), File: "policy.rego", Row: 3, Col: 6},
+				},
+			},
+		},
+		{
+			note: "rule shadows input (multiple modules)",
+			modules: map[string]string{
+				"policy1.rego": `package test
+					import future.strict
+					input := 1`,
+				"policy2.rego": `package test2
+					import future.strict
+					input := 2`,
+			},
+			expectedErrors: Errors{
+				&Error{
+					Message:  "rules must not shadow input (use a different rule name)",
+					Location: &Location{Text: []byte("input := 1"), File: "policy1.rego", Row: 3, Col: 6},
+				},
+				&Error{
+					Message:  "rules must not shadow input (use a different rule name)",
+					Location: &Location{Text: []byte("input := 2"), File: "policy2.rego", Row: 3, Col: 6},
+				},
+			},
+		},
+		{
+			note: "rule shadows input (multiple modules, not all strict)",
+			modules: map[string]string{
+				"policy1.rego": `package test
+					input := 1`,
+				"policy2.rego": `package test2
+					import future.strict
+					input := 2`,
+			},
+			expectedErrors: Errors{
+				&Error{
+					Message:  "rules must not shadow input (use a different rule name)",
+					Location: &Location{Text: []byte("input := 2"), File: "policy2.rego", Row: 3, Col: 6},
+				},
+			},
+		},
+		{
+			note: "rule shadows data",
+			modules: map[string]string{
```

Comment:
What about a test case where a ref-head shadows data/input ?

Reply:
Currently, something like `data.a.b := 1` or `input.a.b := 1` doesn't generate errors. They probably should, though ..

The `input` case will cause a complete replacement of the input document: https://play.openpolicyagent.org/p/ihpFa7iM1p.

The `data` case doesn't produce error or output (!): https://play.openpolicyagent.org/p/shSDlF4mPE

I think it's out of scope to be fixed here, but we should plug these holes.

Reply:
https://github.com/open-policy-agent/opa/issues/6291